lib/SILOptimizer/Mandatory/AddressLowering.cpp

```diff
@@ -690,6 +690,11 @@ static unsigned insertIndirectReturnArgs(AddressLoweringState &pass) {
   unsigned argIdx = 0;
   for (auto resultTy : pass.loweredFnConv.getIndirectSILResultTypes(typeCtx)) {
     auto bodyResultTy = pass.function->mapTypeIntoContext(resultTy);
+    bodyResultTy = pass.function->getModule().Types.getLoweredType(
+        substOpaqueTypesWithUnderlyingTypes(bodyResultTy.getASTType(),
+                                            *pass.function,
+                                            /*allowLoweredTypes=*/true),
```

Comment:
It seems error-prone that we need to call `substOpaqueTypesWithUnderlyingTypes` here.

For anyone following along..
The code above is used to insert the `SILFunctionArgument` for the indirect `@out` function result.

So given:
`sil [ossa] @f076_opaqueResult : $@convention(thin) <T> () -> @out @_opaqueReturnTypeOf("$s5Swift4getSQryF", 0) opaque`

We need to generate:
`bb0(%0 : $*S):`

`getIndirectSILResultTypes` currently returns the opaque type rather than the concrete type `S` even though our type expansion context is the same module as the body of the function. It seems to me like the bug is at a deeper layer.

Reply:
@aschwaighofer @jckarter Do we expect opaque types to be lowered into concrete types during SIL function type lowering? Or should it happen during `mapTypeIntoContext`?

If the function has any other generic arguments, then SIL function type lowering seems to be refusing to lower the opaque result. And mapTypeIntoContext is not doing the type substitution.

Reply:
In `SILGenProlog`'s `emitIndirectResultParameters`, the type used for the out parameter is produced like so:
```
  Type resultType;
  AbstractionPattern origResultType;
  DeclContext *DC;
  CanType resultTypeInContext =
    DC->mapTypeIntoContext(resultType)->getCanonicalType();

  auto &resultTI =
    SGF.SGM.Types.getTypeLowering(origResultType, resultTypeInContext,
                                  SGF.getTypeExpansionContext());
  SILType resultSILType = resultTI.getLoweredType().getAddressType();
```
An attempt at adapting that to the current context yields
```
  for (auto resultTy : pass.loweredFnConv.getIndirectSILResultTypes(typeCtx)) {
    auto resultTyInContext = pass.function->mapTypeIntoContext(resultTy);
    auto bodyResultTy = pass.function->getModule().Types.getLoweredType(
      resultTyInContext.getASTType(),
      *pass.function
    );
```
That give `S`.

In other words, roughly dropping the call to `substOpaqueTypesWithUnderlyingTypes` from this patch (but keeping the call to `getLoweredType`) gets the answer we need.  Is this the right, though?

Reply:
I see. So lowering the function type is insufficient to lower the result type? Strange that it works differently when the function has an unrelated generic parameter.

If we need to separately lower the result type, it seems that should be done in `getIndirectSILResultTypes`.

Anyway, I think we want to map the result into this context *after* lowering the opaque type.

Reply:
> Strange that it works differently when the function has an unrelated generic parameter.

FWIW, the generic parameter isn't _entirely_ unrelated.  Here's what the opaque result of `func vend<T>(_ t: T.Type) -> some P { return S() }` looks like:

```
(opaque_type address=0x1231ef180
  (interface_type=generic_type_param_type depth=1 index=0) conforms_to=main.(file).P@main.swift:2:10 decl=main.(file).vend@main.swift:1:6
  (substitution_map generic_signature=<τ_0_0>
    (substitution τ_0_0 -> τ_0_0)))
```

Presumably that parameterization is to support using the generic parameters in the concrete type like `func vend<T>(_ t: T.Type) -> some P { return S<T>() }`

> If we need to separately lower the result type, it seems that should be done in getIndirectSILResultTypes.

Makes sense.

> we want to map the result into this context after lowering the opaque type.

That's different from what is done in `SILGenProlog`'s `emitIndirectResultParameters`, right?

Reply:
> That's different from what is done in SILGenProlog's emitIndirectResultParameters, right?

Yep. I guess we should be consistent. Maybe that works out because the opaque type has the substitution map. I'm still entirely guessing at how to use these basic type system APIs.

Reply:
Pushed the version described above.  At the moment, it's still doing the lowering separately outside of and after the call to `getIndirectSILResultsTypes`.  FWIW, this same series of calls appears in `SILGenFunction::collectThunkParams`:

```
// lib/SILGen/SILGenPoly.cpp:810
  for (auto resultTy : F.getConventions().getIndirectSILResultTypes(
           getTypeExpansionContext())) {
    auto paramTy = F.mapTypeIntoContext(resultTy);
    // Lower result parameters in the context of the function: opaque result
    // types will be lowered to their underlying type if allowed by resilience.
    auto inContextParamTy = F.getLoweredType(paramTy.getASTType())
                                .getCategoryType(paramTy.getCategory());
```

Reply:
well, SILGen is also grabbing the original ASTType and adjusting the "category", whatever that does. I don't think I'm qualified to say whether this is correct.

Reply:
That's being done already in ``AddressLowering`'s `insertIndirectReturnArgs` via `getAddressType` which returns a SILType whose value category is address:

```
    SILFunctionArgument *funcArg =
        pass.function->begin()->insertFunctionArgument(
            argIdx, bodyResultTy.getAddressType(), OwnershipKind::None, var);
```

Reply:
I'm afraid that SILResultTypes will be used in other places that don't run on all the SIL, so we're unlikely to catch these bugs.

But for the purpose of this PR it seems ok

Reply:
Filed rdar://113770113 .